{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "import Bio\n",
    "from Bio.Seq import Seq\n",
    "from Bio import SeqIO\n",
    "from Bio.SeqRecord import SeqRecord\n",
    "from Bio.SeqFeature import SeqFeature, FeatureLocation\n",
    "from Bio import pairwise2\n",
    "from Bio.pairwise2 import format_alignment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "#region_with_L1HS = SeqIO.read('ref_region_with_L1HS.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1_start_pos = 73002 #note I subtracted one from both to account for Boissinot's likely use of base 1 indexing\n",
    "#L1_end_pos = 79019"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1asfeature = SeqFeature(FeatureLocation(L1_start_pos, L1_end_pos), type='gene', strand = -1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS = L1asfeature.extract(region_with_L1HS)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SeqRecord(seq=Seq('TTATACTCTAAGTTTTAGGGTACATGTGCACATTGTGCAGGTTAGTTACATATG...CCC'), id='<unknown id>', name='<unknown name>', description='<unknown description>', dbxrefs=[])"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#L1HS"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS.id = 'L1PA1_in_AC004554'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#diagnostic_nucleotide_feature = SeqFeature(location=FeatureLocation(start = 5955, end = 5956), type='snp', id = 'diagnostic_nt' )"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS.features.append(diagnostic_nucleotide_feature)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS.features"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS.features[0].extract(L1HS.seq)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "#L1HS.seq[5950:5960]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [],
   "source": [
    "L1HS3prime = Bio.SeqIO.read('L1HS_3prime_DFAM.fa', 'fasta')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "L1HS3prime = L1HS3prime.upper()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_ta_seq = Seq('ACGTTAGTGGGTGC') #see boissinot 2000 fig 4\n",
    "ta_seq = Seq('ACATTAGTGGGTGC')\n",
    "downstream_diagnostic_nuc_seq = Seq('TTAGTGGGTGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "non_ta_2 = Seq(\"GATGACGAG\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "metadata": {},
   "outputs": [],
   "source": [
    "test_seq = Seq(\"CCTAATGCTAGATGACACA\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 183,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def GetDiagnosticNuc(read_seq, target_seq = 'TTAGTGGGTGC'):\n",
    "    \"\"\"\n",
    "    input:\n",
    "    target must be a Seq (by default it is 11 nuc 3' of diagnostic nuc in boissinot), read must be a Seq or string\n",
    "    \n",
    "    output:\n",
    "    returns one of the following lists:\n",
    "    ['likely_ta', read_seq[ACpos:diagnostic_nuc_read_pos+1]\n",
    "    ['likely_non_ta', read_seq[ACpos:diagnostic_nuc_read_pos+1]\n",
    "    ['no_diagnostic_info_in_read']\n",
    "    \"\"\"\n",
    "    #import and define aligner parameters\n",
    "    from Bio import Align\n",
    "\n",
    "    aligner = Align.PairwiseAligner()\n",
    "    aligner.mode = 'local'\n",
    "    aligner.match_score = 2\n",
    "    aligner.mismatch_score = -1\n",
    "    aligner.open_gap_score = -0.5\n",
    "    aligner.extend_gap_score = -0.1\n",
    "    aligner.target_end_gap_score = 0.0\n",
    "    aligner.query_end_gap_score = 0.0\n",
    "\n",
    "    #retrieve diagnostic nuc in reads which have a perfect\n",
    "    #alignment to the sequence immediately downstream of diag nuc\n",
    "\n",
    "    ta_supporting_read = []\n",
    "    non_ta_supporting_read = []\n",
    "    no_info_read = []\n",
    "\n",
    "\n",
    "    for alignment in aligner.align(target_seq, read_seq):\n",
    "        print(\"Score = %.1f:\" % alignment.score)\n",
    "        print(alignment)\n",
    "        print(alignment.aligned)\n",
    "        if alignment.score ==22:\n",
    "        #this means the read has a perfect match to region right after diag nuc\n",
    "        #perfect match requirement may be too stringent\n",
    "            diagnostic_nuc_read_pos = alignment.aligned[1][0][0] - 1\n",
    "            ACpos = alignment.aligned[1][0][0] - 3\n",
    "            if ACpos >=0:\n",
    "                ACseq = read_seq[ACpos:diagnostic_nuc_read_pos]\n",
    "                diagnostic_nuc_read = read_seq[diagnostic_nuc_read_pos]\n",
    "                print(ACseq+diagnostic_nuc_read)\n",
    "                if ACseq == 'AC':\n",
    "                    if diagnostic_nuc_read == 'A':\n",
    "                        return ['likely_ta', read_seq[ACpos:diagnostic_nuc_read_pos+1]]\n",
    "                    elif diagnostic_nuc_read == 'G':\n",
    "                        return ['likely_non_ta', read_seq[ACpos:diagnostic_nuc_read_pos+1]]\n",
    "                    else:\n",
    "                        return ['diagnostic nuc not A or G']\n",
    "                else:\n",
    "                    return ['AC nucleotides absent']\n",
    "                \n",
    "            else: \n",
    "                return [\"\"\"no_diagnostic_info_in_read; read does not extend far enough to see ACA/G\"\"\"]\n",
    "        else:\n",
    "            return ['No perfect alignment with target seq in read']\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 185,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Score = 19.0:\n",
      "   TTAGTGGGTGC\n",
      "   |||||||||.|\n",
      "ACATTAGTGGGTAC\n",
      "\n",
      "(((0, 11),), ((3, 14),))\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "['No perfect alignment with target seq in read']"
      ]
     },
     "execution_count": 185,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "GetDiagnosticNuc('ACATTAGTGGGTGC')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "interpreter": {
   "hash": "48d5d1ce2a3db6f239c4d01dd12b62baeb30c7a9dd3abf83cffd92f51c9167c4"
  },
  "kernelspec": {
   "display_name": "Python 3.8.5 64-bit ('base': conda)",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.5"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}